tools/build_defs/repo/cache.bzl

```diff
@@ -0,0 +1,41 @@
+# Copyright 2023 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# WARNING:
+# https://github.com/bazelbuild/bazel/issues/17713
+# .bzl files in this package (tools/build_defs/repo) are evaluated
+# in a Starlark environment without "@_builtins" injection, and must not refer
+# to symbols associated with build/workspace .bzl files
+
+visibility("private")
+
+NO_DEFAULT_CANONICAL_ID_ENV = "BAZEL_NO_DEFAULT_CANONICAL_ID"
+
+CANONICAL_ID_DOC = """A canonical ID of the file downloaded.
+
+If specified and non-empty, Bazel will not take the file from cache, unless it
+was added to the cache by a request with the same canonical ID.
+
+If unspecified or empty, Bazel by default uses the URLs of the file as the
+canonical ID. This helps catch the common mistake of updating the URLs without
+also updating the hash, resulting in builds that succeed locally but fail on
+machines without the file in the cache. This behavior can be disabled with
+--repo_env={env}=1.
+""".format(env = NO_DEFAULT_CANONICAL_ID_ENV)
+
+def get_default_canonical_id(repository_ctx, urls):
+    """Returns the default canonical id to use for downloads."""
+    if repository_ctx.os.environ.get(NO_DEFAULT_CANONICAL_ID_ENV) == "1":
+        return ""
+    return " ".join(urls)
```

Comment:
Nit: might want to sort this before join

Reply:
This actually isn't safe, will change and add a comment.

Reply:
Added, please take a look.

Reply:
Hmm strange… I expected this exact “broken” behavior 🤔 . Theoretically if A is broken, Bazel will retry on B anyway so cache hit should be ok? I cannot think of a scenario where I would want different cache entries when changing the URLs order.

I guess this is something that the indirection we discussed could come into play?

Reply:
If A is broken in the sense of returning 404 or being unresponsive, then Bazel will fail over to B. But if A happily returns a file with a hash or prefix that doesn't match, then Bazel will fail immediately. That's why I think order does matter.

In fact the indirection is already implemented: There is only ever one repository cache entry per hash and this single entry additionally tracks all associated canonical IDs.